chunky/src/java/se/llbit/math/BVH.java

```diff
@@ -17,530 +17,63 @@
-      } else {
-        Primitive[] chunk = chunks.pop();
-        if (chunk.length < SPLIT_LIMIT) {
-          nodes.push(new Leaf(chunk));
-        } else {
-          splitSAH_MA(chunk, actions, chunks);
-        }
-      }
-    }
-    return nodes.pop();
-  }
+  public final BVHImplementation implementation;
 
   /**
-   * Split a chunk based on Surface Area Heuristic of all possible splits
+   * Construct a new BVH containing the given entities. This will generate the BVH using the
+   * persistent BVH method (default is SAH_MA).
    */
-  private void splitSAH(Primitive[] chunk, Stack<Action> actions, Stack<Primitive[]> chunks) {
-    MutableAABB bounds = new MutableAABB(0, 0, 0, 0, 0, 0);
-    double cmin = Double.POSITIVE_INFINITY;
-    int split = 0;
-    int end = chunk.length;
-
-    double[] sl = new double[end];
-    double[] sr = new double[end];
-
-    Comparator<Primitive> cmp = cmpX;
-    Arrays.sort(chunk, cmpX);
-    for (int i = 0; i < end - 1; ++i) {
-      bounds.expand(chunk[i].bounds());
-      sl[i] = bounds.surfaceArea();
-    }
-    bounds = new MutableAABB(0, 0, 0, 0, 0, 0);
-    for (int i = end - 1; i > 0; --i) {
-      bounds.expand(chunk[i].bounds());
-      sr[i - 1] = bounds.surfaceArea();
-    }
-    for (int i = 0; i < end - 1; ++i) {
-      double c = sl[i] * (i + 1) + sr[i] * (end - i - 1);
-      if (c < cmin) {
-        cmin = c;
-        split = i;
-      }
-    }
-
-    Arrays.sort(chunk, cmpY);
-    for (int i = 0; i < end - 1; ++i) {
-      bounds.expand(chunk[i].bounds());
-      sl[i] = bounds.surfaceArea();
-    }
-    bounds = new MutableAABB(0, 0, 0, 0, 0, 0);
-    for (int i = end - 1; i > 0; --i) {
-      bounds.expand(chunk[i].bounds());
-      sr[i - 1] = bounds.surfaceArea();
-    }
-    for (int i = 0; i < end - 1; ++i) {
-      double c = sl[i] * (i + 1) + sr[i] * (end - i - 1);
-      if (c < cmin) {
-        cmin = c;
-        split = i;
-        cmp = cmpY;
-      }
-    }
-
-    Arrays.sort(chunk, cmpZ);
-    for (int i = 0; i < end - 1; ++i) {
-      bounds.expand(chunk[i].bounds());
-      sl[i] = bounds.surfaceArea();
-    }
-    bounds = new MutableAABB(0, 0, 0, 0, 0, 0);
-    for (int i = end - 1; i > 0; --i) {
-      bounds.expand(chunk[i].bounds());
-      sr[i - 1] = bounds.surfaceArea();
-    }
-    for (int i = 0; i < end - 1; ++i) {
-      double c = sl[i] * (i + 1) + sr[i] * (end - i - 1);
-      if (c < cmin) {
-        cmin = c;
-        split = i;
-        cmp = cmpZ;
-      }
-    }
-
-    if (cmp != cmpZ) {
-      Arrays.sort(chunk, cmp);
-    }
-
-    split += 1;
-
-    actions.push(Action.MERGE);
-    Primitive[] cons = new Primitive[split];
-    System.arraycopy(chunk, 0, cons, 0, split);
-    chunks.push(cons);
-    actions.push(Action.PUSH);
-
-    cons = new Primitive[end - split];
-    System.arraycopy(chunk, split, cons, 0, end - split);
-    chunks.push(cons);
-    actions.push(Action.PUSH);
+  public BVH(Collection<Entity> entities, Vector3 worldOffset) {
```

Comment:
The BVH doesn't really need the `Entity` interface, it should work fine with any `Primitive`, as before. Seemed like a better level of abstraction.

Reply:
This is to add future support for entity based BVH's which should be much more memory efficient.

Reply:
…or (if the intention is not to create that `Collection<Primitive>` for every entity) maybe a `PrimitiveModel` interface (that is implemented by `Entity`) that has a `getPrimitives(Consumer<Primitive> addPrimitive)` method?

Reply:
@alexhliu What's an entity-based bvh? In the end we'll always pathtrace primitives… there is no simple intersection function for "Creeper shape"

Reply:
One idea was to make a two level BVH. The top level is a normal BVH around the entities. The second level are BVH's for each unique entity and deduplicated where possible. Basically a DICTIONARY octree but in BVH form.